pkg/logging/logging.go

```diff
@@ -0,0 +1,45 @@
+package logging
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/felixge/httpsnoop"
+	log "github.com/sirupsen/logrus"
+)
+
+func New() func(http.Handler) http.Handler {
```

Comment:
This doesn't allow for any real middleware setup or options passing. I think we need to find a standard to conform to that is flexible enough for future additions to this repo.

Currently, I see two possibilities we could go for. A typed approach that allows for flexibility in setting up complex middleware, and another that is more concise with little flexibility.
```
package example

import (
  "net/http"
)

// middleware as custom type with constructor
type Typed struct{}

func NewTyped(opts ...func(*Typed)) *Typed {
  m := &Typed{}

  for _, opt := range opts {
    opt(m)
  }

  return m
}

func (m *Typed) Middleware(next http.Handler) http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    // do something
  })
}

// middleware as function
func NewFunced(next http.Handler) http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    // do something
  })
}
```

These can be easily used with either `http.ServeMux` or the gorilla/mux style of middleware:
```
m := NewTyped(opts...)

muxRouter := mux.NewRouter()
// handler middleware
muxRouter.HandleFunc("/", m.Middleware(someHandler))

// global middleware
muxRouter.Use(m.Middleware)

serveMux := http.NewServeMux()
// handler middleware
serveMux.HandleFunc("/", m.Middleware(someHandler))

// global middleware
m.Middleware(serveMux)
```

For the sake of simplicity, I think we could probably just go with the typed approach for all middleware.

Reply:
I do think the middleware should be unnecessarily typed.  Both  http.ServeMux and gorilla/mux support the current return type.  

If a function is defined as `func New (opts ...Option)`  the `New()` and `New(opt1, opt2, opt3)`  are both valid.  Even in the case a middleware needs to in the future extend types, it can do so without breaking compatibility. 

I also do not like returning a `*Typed since that can open up the possibility to re-configure middleware behavior after the fact.  

I do not understand why two separate functions would need to be offered when they are both equivalent to this return type `func(http.Handler) http.Handler`

Reply:
> I do think the middleware should be unnecessarily typed. Both http.ServeMux and gorilla/mux support the current return type.

There are cases where a type and pointer methods are necessary, like in the [example in the mux readme](https://pkg.go.dev/github.com/gorilla/mux#readme-middleware).

> If a function is defined as `func New (opts ...Option)` the `New()` and `New(opt1, opt2, opt3)` are both valid. Even in the case a middleware needs to in the future extend types, it can do so without breaking compatibility.

Agreed, this just isn't providing a way for options to be stored or passed to the actual middleware function. I'm proposing we use a consistent approach using types and middleware methods on those types now so that this doesn't develop into each middleware handling this is subtly different ways. 

> I also do not like returning a `*Typed since that can open up the possibility to re-configure middleware behavior after the fact.

Not really sure I follow. This seems like more of a user concern in how they setup middleware and routes.

> I do not understand why two separate functions would need to be offered when they are both equivalent to this return type `func(http.Handler) http.Handler`

Two separate functions aren't needed. I was just pointing out that the constructor isn't necessary if we aren't providing a way to parse options. I think a single typed approach for all middleware is probably the better way to go.

Reply:
I am all for defining a standard and will include the necessary documentation for how to create middleware.  In the case a middleware does not require extra configuration options, providing typed structure feels unnecessary.  

This will likely change for `logging` as I add options for logrus.

Reply:
> In the case a middleware does not require extra configuration options, providing typed structure feels unnecessary.

I disagree. At a bare minimum a type provide a convenient place to dangle a middleware method from, and provides an obvious path for parsing options should it become necessary.

Reply:
I'll compromise on returning an Interface provider for the middleware function. Underlying types should be internal and only configurable via the varadic option functions.

Reply:
Added a new comment about the interface. Closing this thread.